src/cascadia/TerminalConnection/ConhostConnection.cpp

```diff
@@ -190,38 +193,75 @@ namespace winrt::Microsoft::Terminal::TerminalConnection::implementation
     DWORD ConhostConnection::_OutputThread()
     {
         const size_t bufferSize = 4096;
-        BYTE buffer[bufferSize];
-        DWORD dwRead;
+        BYTE buffer[bufferSize]{ 0 };
+        BYTE utf8Partials[4]{ 0 }; // buffer for code units of a partial UTF-8 code point that have to be cached
+        DWORD dwRead{}; // length of a chunk read
+        DWORD dwPartialsLen{}; // number of cached UTF-8 code units
+        bool fSuccess{};
+        const BYTE bitmasks[]{ 0, 0b11000000, 0b11100000, 0b11110000 }; // for comparisons after the Lead Byte was found
+
+        // process the data of the output pipe in a loop
         while (true)
         {
-            dwRead = 0;
-            bool fSuccess = false;
+            // copy UTF-8 code units that were remaining from the previously read chunk (if any)
+            if (dwPartialsLen != 0)
+            {
+                std::move(utf8Partials, utf8Partials + dwPartialsLen, buffer);
+            }
+
+            // try to read data
+            fSuccess = !!ReadFile(_outPipe.get(), &buffer[dwPartialsLen], std::extent<decltype(buffer)>::value - dwPartialsLen, &dwRead, nullptr);
+
+            dwRead += dwPartialsLen;
+            dwPartialsLen = 0;
 
-            fSuccess = !!ReadFile(_outPipe.get(), buffer, bufferSize, &dwRead, nullptr);
-            if (!fSuccess)
+            if (dwRead == 0) // quit if no data has been read and no cached data was left over
+            {
+                return 0;
+            }
+            else if (!fSuccess)
             {
                 if (_closing.load())
                 {
                     // This is okay, break out to kill the thread
                     return 0;
                 }
-                else
-                {
-                    _disconnectHandlers();
-                    return (DWORD)-1;
-                }
+
+                _disconnectHandlers();
+                return (DWORD)-1;
             }
-            if (dwRead == 0)
+
+            const BYTE* const endPtr{ buffer + dwRead };
+            const BYTE* backIter{ endPtr - 1 };
+            // If the last byte in the buffer was a byte belonging to a UTF-8 multi-byte character
+            if ((*backIter & 0b10000000) == 0b10000000)

```

Comment:
We should _probably_ be using the wil `WI_AreAllFlagsSet` (et. al.) for these comparisons.

Reply:
Yes that's an option if you like it better. Let me know.

Reply:
Yeah, let's do that!

Reply:
Here it goes.